Implement countSubstrings()

azat opened the pull request:
Changelog category (leave one):
- Improvement

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Implement `countSubstrings()`/`countSubstringsCaseInsensitive()`/`countSubstringsCaseInsensitiveUTF8()` (Count the number of substring occurrences)

Detailed description / Documentation draft:
The code is based on position helpers.

alexey-milovidov commented:
Maybe add a test for empty substring?

alexey-milovidov commented on docs/en/sql-reference/functions/string-search-functions.md:
```diff
@@ -536,4 +536,44 @@ For case-insensitive search or/and in UTF-8 format use functions `ngramSearchCas
+
+Count the number of substring occurrences
+
+For a case-insensitive search, use the function `substringCountCaseInsensitive`.
```
Is there `CaseInsensitiveUTF8` variant?

azat replied:
In the latest version - yes (3265b29c19c2413f5275d8b05271299e718beea0)

alexey-milovidov commented on docs/en/sql-reference/functions/string-search-functions.md:
```diff
@@ -536,4 +536,44 @@ For case-insensitive search or/and in UTF-8 format use functions `ngramSearchCas
+
+**Returned values**
+
+-   Number of occurrences.
```
How to count if substring is empty?

(throw an exception; return the number of chars in haystack; return the number of chars + 1 - all three variants Ok).

alexey-milovidov replied:
No mention about intersecting substrings in haystack.
Example: 

```
substringCount('aaaa', 'aa')
```

may return 2 for non-intersecting substrings:
```
aaaa
aa
  aa
```

or 3 for intersecting substrings:
```
aaaa
aa
 aa
  aa
```

alexey-milovidov replied:
(it's Ok to return non-intersecting substrings only)

alexey-milovidov replied:
Let's also add a test to check this behaviour.

alexey-milovidov replied:
It's better to count non-intersecting substrings, because it will be consistent with the `extractAll` function.

alexey-milovidov replied:
The number of returned substrings for empty needle should be consistent with the returned array size of `extractAll` when empty regexp is passed.

azat replied:
>How to count if substring is empty?

It will return zero (same as `extractAll`, tests updated)

>The number of returned substrings for empty needle should be consistent with the returned array size of extractAll when empty regexp is passed.

Yes, it will also return zero (tests updated)

>It's better to count non-intersecting substrings, because it will be consistent with the extractAll function.

In the latest version this has been addressed (tests updated for all variants and add some comments in code just in case)

alexey-milovidov commented on src/Functions/SubstringCountImpl.h:
```diff
@@ -0,0 +1,221 @@
+        while (pos < end && end != (pos = searcher.search(pos, end - pos)))
+        {
+            /// Determine which index it refers to.
+            while (begin + offsets[i] <= pos) ++i;
```
Style.

azat committed:
Implement countSubstrings()

Function to count number of substring occurrences in the string:
- in case of needle is multi char - counts non-intersecting substrings
- the code is based on position helpers.

The following new functions is available:
- countSubstrings()
- countSubstringsCaseInsensitive()
- countSubstringsCaseInsensitiveUTF8()

v0: substringCount()

v2:
- add substringCountCaseInsensitiveUTF8
- improve tests
- fix coding style issues
- fix multichar needle

v3: rename to countSubstrings (by analogy with countEqual())

alexey-milovidov commented:
Please review (and continue?) this PR https://github.com/ClickHouse/ClickHouse/pull/16277

alexey-milovidov commented:
The function should be named countSubstrings, not substringCount for consistency with the countEquals function.

azat commented:
>Please review (and continue?) this PR #16277

Ok, I will take a look

>The function should be named countSubstrings, not substringCount for consistency with the countEquals function.

Ok, renamed (that PR does not implements it anyway, so I will continue it here)

azat commented:
>Please review (and continue?) this PR #16277

Since the author hadn't responding, I decided to just continue it, see #17459 (all changes was done on top of that PR)
